distributed/comm/tests/test_ucx.py

```diff
@@ -0,0 +1,290 @@
+            msg = await comm.read()
+            msg["op"] = "pong"
+            await comm.write(msg)
+            assert comm.closed() is False
+            await comm.close()
+            assert comm.closed
+
+        listener = ucx.UCXListener(address, handle_comm)
+        listener.start()
+        host, port = listener.get_host_port()
+        assert host.count(".") == 3
+        assert port > 0
+
+        connector = ucx.UCXConnector()
+        l = []
+
+        async def client_communicate(key, delay=0):
+            addr = "%s:%d" % (host, port)
+            comm = await connector.connect(addr)
+            # TODO: peer_address
+            # assert comm.peer_address == 'ucx://' + addr
+            assert comm.extra_info == {}
+            msg = {"op": "ping", "data": key}
+            await comm.write(msg)
+            if delay:
+                await asyncio.sleep(delay)
+            msg = await comm.read()
+            assert msg == {"op": "pong", "data": key}
+            l.append(key)
+            return comm
+            assert comm.closed() is False
+            await comm.close()
+            assert comm.closed
+
+        comm = await client_communicate(key=1234, delay=0.5)
+
+        # Many clients at once
+        N = 2
+        futures = [client_communicate(key=i, delay=0.05) for i in range(N)]
+        await asyncio.gather(*futures)
+        assert set(l) == {1234} | set(range(N))
+
+    asyncio.run(f())
+
+
+@pytest.mark.asyncio
+async def test_ping_pong_data():
+    np = pytest.importorskip("numpy")
+
+    data = np.ones((10, 10))
+    # TODO: broken for large arrays
+    address = "ucx://{}:{}".format(HOST, next(port_counter))
+    com, serv_com = await get_comm_pair(address)
+    msg = {"op": "ping", "data": to_serialize(data)}
+    await com.write(msg)
+    result = await serv_com.read()
+    result["op"] = "pong"
+    data2 = result.pop("data")
+    np.testing.assert_array_equal(data2, data)
+
+    await serv_com.write(result)
+
+    result = await com.read()
+    assert result == {"op": "pong"}
+
+    await com.close()
+    await serv_com.close()
+
+
+@gen_test()
+def test_ucx_deserialize():
+    yield check_deserialize("tcp://")
+
+
+@pytest.mark.asyncio
+async def test_ping_pong_cudf():
+    # if this test appears after cupy an import error arises
+    # *** ImportError: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `CXXABI_1.3.11'
+    # not found (required by python3.7/site-packages/pyarrow/../../../libarrow.so.12)
+    cudf = pytest.importorskip("cudf")
+
+    df = cudf.DataFrame({"A": [1, 2, None], "B": [1.0, 2.0, None]})
+    address = "ucx://{}:{}".format(HOST, next(port_counter))
+
+    com, serv_com = await get_comm_pair(address)
+    msg = {"op": "ping", "data": to_serialize(df)}
+
+    await com.write(msg)
+    result = await serv_com.read()
+    data2 = result.pop("data")
+    assert result["op"] == "ping"
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("shape", [(100,), (10, 10), (4_947,)])
```

Comment:
@pentschev I'm able to get this test to fail (it hangs actually) if I have an array of this size or larger (one fewer element and things are fine).  Does this sound familiar to you?  I've merged in your latest changes from master.

Reply:
It looks like we get stuck here:

https://github.com/dask/distributed/pull/2591/files#diff-ca5c24c584111a291f763107e4065539R168

We finish the `self.ep.send_obj(frame)` call, but then get stuck waiting on the future, which never completes.  @Akshay-Venkatesh, does this sound familiar to you?  Is there some behavior that might change around `4946 * 8 == 39568` bytes?  We send a few frames before the main piece of data.  These have the following lengths:

```
(Pdb) pp list(map(len, frames[:-1]))
[4, 32, 0, 9, 259]
```

Reply:
Just to be clear here, we're getting block on waiting for the *send* to complete, which seems odd to me.

Reply:
The message size sticks out because it would likely be the range where rendezvous protocol is used. So unlike the short message sizes, if the corresponding receive isn't posted, then this send operation blocks until such time. Is the receive being posted?

Reply:
Also, can you send me repro commands for `test_ucx.py`? Matt, looks you're able to get this test to fail (https://github.com/dask/distributed/pull/2591/files#r279561588) so if I can repro I maybe more useful.

Reply:
(thinking about the behavior of recvs in dask, one sec)

In the meantime, this is maybe also relevant.  Here is how this manifests itself in practice.

```python
In [1]: from dask.distributed import Client

In [2]: client = Client(protocol='ucx', interface='ib0')

In [3]: import cupy

In [4]: x = cupy.random.randint(0, 255, size=30000, dtype='uint8')

In [5]: client.scatter(x).result()  # roundtrip
Out[5]: array([152, 251, 139, ...,   4, 136, 137], dtype=uint8)

In [6]: x = cupy.random.randint(0, 255, size=40000, dtype='uint8')

In [7]: client.scatter(x).result()  # roundtrip
[dgx15:72708:0:72858] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x28)
==== backtrace ====
    0  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucs.so.0(+0x21b50) [0x7fce860f4b50]
    1  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucs.so.0(+0x21dc4) [0x7fce860f4dc4]
    2  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucp.so.0(ucp_rkey_destroy+0x7) [0x7fce867f8b27]
    3  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucp.so.0(ucp_rndv_progress_rma_get_zcopy+0x445) [0x7fce86814375]
    4  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucp.so.0(ucp_rndv_rtr_handler+0x65d) [0x7fce8681880d]
    5  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libuct.so.0(uct_rc_mlx5_iface_progress+0x11e) [0x7fce8657146e]
    6  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucp.so.0(ucp_worker_progress+0x4a) [0x7fce867fde8a]
    7  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0xace4c) [0x7fce86b01e4c]
    8  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x73fb5) [0x7fce86ac8fb5]
    9  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x25e9d) [0x7fce86a7ae9d]
   10  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x477a3) [0x7fce86a9c7a3]
   11  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(PyCFunction_Call+0x65) [0x55baae8efb65]
   12  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x1eb70) [0x7fce86a73b70]
   13  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x46dc3) [0x7fce86a9bdc3]
   14  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyMethodDef_RawFastCallKeywords+0x305) [0x55baae92c785]
   15  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyCFunction_FastCallKeywords+0x21) [0x55baae92c801]
   16  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyEval_EvalFrameDefault+0x544e) [0x55baae98887e]
   17  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyGen_Send+0x14c) [0x55baae92551c]
   18  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/python3.7/lib-dynload/_asyncio.cpython-37m-x86_64-linux-gnu.so(+0xabc9) [0x7fcea9a24bc9]
   19  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyObject_FastCallKeywords+0x49b) [0x55baae92cd2b]
   20  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(+0x20abc3) [0x55baae9bdbc3]
   21  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyMethodDef_RawFastCallDict+0x199) [0x55baae8ef8b9]
   22  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyCFunction_FastCallDict+0x21) [0x55baae8efae1]
   23  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyEval_EvalFrameDefault+0x5d21) [0x55baae989151]
   24  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyFunction_FastCallKeywords+0xfb) [0x55baae92b79b]
   25  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyEval_EvalFrameDefault+0x6a0) [0x55baae983ad0]
```

Reply:
Right, so in that test if I apply the following diff then things no longer hang, instead I get a segfault like what is shown above (in practice, there is always a pending recv).

```diff

-    await com.write(msg)
-    result = await serv_com.read()
+    _, result = await asyncio.gather(com.write(msg), serv_com.read())
```

Reply:
@mrocklin for the `rkey_destroy` error, can you run the experiment by setting `UCX_MEMTYPE_CACHE=n UCX_RNDV_SCHEME=put_zcopy` env vars (in case you've not already). We've seen this before.

Reply:
Hooray!

```python
In [1]: from dask.distributed import Client
In [2]: client = Client(protocol='ucx', interface='ib0')

In [3]: import cupy
In [4]: x = cupy.random.randint(0, 255, size=10000000, dtype='uint8')

In [5]: client.scatter(x).result()  # roundtrip
Out[5]: array([166, 159,  69, ..., 226,   2,   0], dtype=uint8)
```

OK, how should we handle this for end users?  Is this something that should be fixed upstream somewhere?  Is this where https://github.com/Akshay-Venkatesh/ucx-py/issues/83 comes in?

Reply:
Also, it looks like there might be a size limit on passed frames?  (if so, we can happily handle this on our end).

```python
In [1]: from dask.distributed import Client

In [2]: client = Client(protocol='ucx', interface='ib0', n_workers=1)

In [3]: import cupy

In [4]: x = cupy.random.randint(0, 255, size=100_000_000, dtype='uint8')

In [5]: %timeit client.scatter(x, hash=False, direct=True).result()
1.26 s ± 7.58 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: x = cupy.random.randint(0, 255, size=1_000_000_000, dtype='uint8')

In [7]: %timeit client.scatter(x, hash=False, direct=True).result()
[1556581592.967171] [dgx15:81863:0]         ucp_mm.c:108  UCX  ERROR failed to register address (nil) length 1000000000 on md[2]=ib/mlx5_3: Invalid parameter
[1556581592.967206] [dgx15:81863:0]    ucp_request.c:263  UCX  ERROR failed to register user buffer datatype 0x8 address (nil) len 1000000000: Invalid parameter
[dgx15:81863:0:595] Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))
==== backtrace ====
    0  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucs.so.0(+0x21b50) [0x7fe96f58ab50]
    1  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucs.so.0(+0x21dc4) [0x7fe96f58adc4]
    2  /lib/x86_64-linux-gnu/libc.so.6(+0x14e156) [0x7fe9816e6156]
    3  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucp.so.0(ucp_rndv_data_handler+0xf3) [0x7fe96fcaec83]
    4  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libuct.so.0(uct_rc_mlx5_iface_progress+0x11e) [0x7fe96fa0746e]
    5  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucp.so.0(ucp_worker_progress+0x4a) [0x7fe96fc93e8a]
    6  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0xace4c) [0x7fe96ff97e4c]
    7  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x6dad9) [0x7fe96ff58ad9]
    8  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyMethodDef_RawFastCallKeywords+0x141) [0x5592534465c1]
    9  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyCFunction_FastCallKeywords+0x21) [0x559253446801]
   10  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(+0x20abc3) [0x5592534d7bc3]
   11  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyMethodDef_RawFastCallDict+0x199) [0x5592534098b9]
   12  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyCFunction_FastCallDict+0x21) [0x559253409ae1]
   13  /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python(_PyEval_EvalFrameDefault+0x5d21) [0x5592534a3151]
```

Also, now I really want NVLINK for ucx-py :)

Reply:
It's a combination of both, Matt. The immediate action would be to set this env var by default such that whenever `ucp` module is loaded we export these env vars.  Long term is to get this fixed in UCX C library so that we don't have to set the env var explicitly. This is because `get` rndv protocol can yield much better overlap.

Reply:
NVLINK PR has stalled recently. Will poll on the UCX devs now.

Reply:
Looks like you've already figured out what was the issue here, it isn't related to the deserialization PR from yesterday, is that right @mrocklin?

Reply:
Correct.  This was something different.

Reply:
Results on cupy arrays.  I get something like 3GB/s.

```python
from bench_comm import run_bench
import asyncio
import cupy

x = cupy.random.randint(0, 255, 200_000_000, dtype='uint8')

loop = asyncio.get_event_loop()
%prun loop.run_until_complete(run_bench('ucx', 0, 50, data=x))
```

```
         34155 function calls (34133 primitive calls) in 3.480 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      506    2.138    0.004    2.138    0.004 {method 'poll' of 'select.epoll' objects}
      301    0.719    0.002    0.720    0.002 {method 'recv_obj' of 'ucp_py.Endpoint' objects}
        1    0.450    0.450    0.450    0.450 {ucp_py.init}
        1    0.057    0.057    0.057    0.057 {ucp_py.get_endpoint}
      503    0.025    0.000    0.026    0.000 {method 'send_obj' of 'ucp_py.Endpoint' objects}
      506    0.008    0.000    3.479    0.007 base_events.py:1679(_run_once)
      824    0.007    0.000    1.323    0.002 {method 'run' of 'Context' objects}
      302    0.004    0.000    0.742    0.002 ucx.py:177(read)
      506    0.004    0.000    2.143    0.004 selectors.py:451(select)
      513    0.003    0.000    0.005    0.000 base_events.py:707(_call_soon)
      202    0.003    0.000    0.003    0.000 {method 'recv_future' of 'ucp_py.Endpoint' objects}
      301    0.003    0.000    0.047    0.000 ucx.py:154(write)
      201    0.002    0.000    0.002    0.000 {msgpack._cmsgpack.unpackb}
      203    0.002    0.000    0.026    0.000 gen.py:184(wrapper)
      202    0.002    0.000    0.002    0.000 {method 'recv' of '_socket.socket' objects}
      101    0.002    0.000    0.006    0.000 core.py:111(loads)
      824    0.002    0.000    1.325    0.002 events.py:86(_run)
```

Reply:
Sometimes I run into this issue:

```
[1556630301.486209] [dgx15:8953 :0]         ucp_mm.c:108  UCX  ERROR failed to register address (nil) length 200000000 on md[5]=ib/mlx5_0: Invalid parameter
[1556630301.486234] [dgx15:8953 :0]    ucp_request.c:263  UCX  ERROR failed to register user buffer datatype 0x8 address (nil) len 200000000: Invalid parameter
[dgx15:8953 :0:8953] Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))
==== backtrace ====
    0  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucs.so.0(+0x21b50) [0x7fd635615b50]
    1  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucs.so.0(+0x21dc4) [0x7fd635615dc4]
    2  /lib/x86_64-linux-gnu/libc.so.6(+0x14e156) [0x7fd644873156]
    3  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucp.so.0(ucp_rndv_data_handler+0xf3) [0x7fd635d39c83]
    4  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libuct.so.0(uct_rc_mlx5_iface_progress+0x11e) [0x7fd635a9246e]
    5  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucp.so.0(ucp_worker_progress+0x4a) [0x7fd635d1ee8a]
    6  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0xace4c) [0x7fd636022e4c]
    7  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x73fb5) [0x7fd635fe9fb5]
    8  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x25e9d) [0x7fd635f9be9d]
    9  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x477a3) [0x7fd635fbd7a3]
   10  python(PyCFunction_Call+0x65) [0x55fb22b88b65]
   11  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x1eb70) [0x7fd635f94b70]
   12  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x46dc3) [0x7fd635fbcdc3]
   13  python(_PyMethodDef_RawFastCallKeywords+0x305) [0x55fb22bc5785]
   14  python(_PyCFunction_FastCallKeywords+0x21) [0x55fb22bc5801]
   15  python(_PyEval_EvalFrameDefault+0x544e) [0x55fb22c2187e]
```

Is this the result of sending around frames that are too large?  Any thoughts @Akshay-Venkatesh?